### PR TITLE
feat(druid): add Apache Druid backend

### LINKED ISSUE #5607: feat: Apache Druid backend
### Is your feature request related to a problem?

_No response_

### Describe the solution you'd like

I'd like to use ibis for working with data in Apache Druid. Are there any plans to implement an ibis backend for it?

### What version of ibis are you running?

4.1.0

### What backend(s) are you using, if any?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by cpcloud**
This PR adds an experimental [Apache Druid](https://druid.apache.org/) backend. It is a SQLAlchemy-based backend by way of [pydruid](https://github.com/druid-io/pydruid).

Closes #5607.

### CODE REVIEW

**Comment by cpcloud** [comment]
cc @sergei3000

File: ibis/backends/druid/__init__.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            warnings.filterwarnings(
+                "ignore",
+                message="|".join(
+                    (
+                        "Did not recognize type",
+                        "Dialect druid:rest will not make use of SQL compilation caching",
+                    )
+                ),
+                category=sa.exc.SAWarning,
+            )
```
Comment (gforsyth) [inline ibis/backends/druid/__init__.py:101]: we could put a `super` call inside the `catch_warnings` context manager and keep the table selection business in `sql/alchemy/__init__.py`

Comment (cpcloud) [reply ibis/backends/druid/__init__.py]: Sweet, thank you. That is much better!

File: docker/druid/environment [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+# under the License.
+#
+
+# Java tuning
+#DRUID_XMX=1g
+#DRUID_XMS=1g
+#DRUID_MAXNEWSIZE=250m
+#DRUID_NEWSIZE=250m
+#DRUID_MAXDIRECTMEMORYSIZE=6172m
+DRUID_SINGLE_NODE_CONF=micro-quickstart
```
Comment (gforsyth) [inline docker/druid/environment:26]:
```suggestion
DRUID_SINGLE_NODE_CONF=nano-quickstart
```

The `micro-quickstart` assumes 4 cores and 16GB of ram, which I think is what is killing the github action worker.

Comment (cpcloud) [reply docker/druid/environment]: Let's give it a shot!

**Comment by kszucs** [comment]
Hmm, the tests just hang for me, probably trying to connect to druid without success. Need to investigate the docker setup in details.

**Comment by cpcloud** [comment]
It's probably a worker capacity issue, and not related to connecting.

If you don't have enough workers the ingestion process just hangs, it's really frustrating.

I personally don't think I want to spend time debugging this for macOS, but since it's working in CI, hanging on macOS doesn't seem like a blocker.

**Comment by kszucs** [comment]
> I personally don't think I want to spend time debugging this for macOS, but since it's working in CI, hanging on macOS doesn't seem like a blocker.

Sure, I just want to be able to run the druid tests. How many workers do I need?

**Comment by cpcloud** [comment]
> > I personally don't think I want to spend time debugging this for macOS, but since it's working in CI, hanging on macOS doesn't seem like a blocker.
> 
> Sure, I just want to be able to run the druid tests. How many workers do I need? `druid_worker_capacity=6`?
> 
> 

Yeah try bumping up 1 or 2 workers.

Also good to confirm it's not a connectivity issue by curling the query endpoint, which should be in the main backend init file.

**Comment by gforsyth** [comment]
> Yeah try bumping up 1 or 2 workers.

Also have `btop` or similar up and ready the first time you try this or you'll lock up your machine when druid eats all your RAM

**Pull request merged**